File: profiles/tom/hooks/logcompress-hook.js
#!/usr/bin/env node
// Adapted from https://github.com/Green-PT/honey-for-devs hooks/ (MIT)
// PostToolUse hook: collapse repetitive Bash output before it lands in context,
// and stash the original so any detail stays retrievable.
//
// Always-on (no mode gate). Two emit conditions:
//   • ≥1 lines were collapsed (repeated-run dedup)
//   • ANSI stripping alone saved >200 chars (pure token noise)
// Falls through silently on small / already-clean output. Fail-open everywhere:
// any throw → exit 0 → original result reaches the model unchanged.
"use strict";
const fs = require("fs");
const os = require("os");
const path = require("path");
const crypto = require("crypto");
const { spawnSync } = require("child_process");

function emit(obj) { process.stdout.write(JSON.stringify(obj)); }
function passthrough() { process.exit(0); } // no output → original result is kept

// Stay well inside the provider's 10s hook deadline. The child owns parsing,
// compression, and cache I/O; the supervisor emits only a complete, validated
// JSON envelope. A slow filesystem or pathological payload is therefore a
// clean no-op, never a provider timeout or a partial replacement envelope.
if (process.env.LOGCOMPRESS_INNER !== "1") {
  try {
    const inputRead = spawnSync(
      "timeout",
      ["--signal=TERM", "--kill-after=0.1s", "1s", "cat"],
      {
        stdio: [0, "pipe", "ignore"],
        maxBuffer: 64 * 1024 * 1024,
        timeout: 2000,
        killSignal: "SIGKILL",
      },
    );
    if (inputRead.status !== 0) process.exit(0);
    const input = inputRead.stdout;
    const dial = spawnSync(
      "bash",
      [
        "-c",
        'source "$1" && north_hook_enabled "$2"',
        "logcompress-dial",
        path.join(__dirname, "lib", "harness-dial.sh"),
        "logcompress-hook",
      ],
      {
        stdio: ["ignore", "ignore", "ignore"],
        timeout: 1000,
        killSignal: "SIGKILL",
      },
    );
    // Status 1 is the resolver's deliberate off verdict. Any other failure is
    // fail-open so telemetry plumbing cannot erase the original tool result.
    if (dial.status === 1) process.exit(0);
    const maxBuffer = Math.min(
      Math.max(input.length * 2 + 1024 * 1024, 1024 * 1024),
      64 * 1024 * 1024,
    );
    const child = spawnSync(
      "timeout",
      ["--signal=TERM", "--kill-after=0.2s", "2s", process.execPath, __filename],
      {
        input,
        env: { ...process.env, LOGCOMPRESS_INNER: "1" },
        encoding: "utf8",
        maxBuffer,
        timeout: 3000,
        killSignal: "SIGKILL",
      },
    );
    if (child.status === 0 && child.stdout) {
      const payload = JSON.parse(child.stdout);
      if (payload?.hookSpecificOutput?.hookEventName === "PostToolUse") {
        process.stdout.write(child.stdout);
      }
    }
  } catch {
    // No output preserves the provider's original tool result.
  }
  process.exit(0);
}

try {
  if (process.env.LOGCOMPRESS_TEST_HANG === "1") {
    if (process.env.LOGCOMPRESS_TEST_PID_FILE) {
      fs.writeFileSync(process.env.LOGCOMPRESS_TEST_PID_FILE, String(process.pid));
    }
    Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 30000);
  }
  const { compress } = require(path.join(__dirname, "logcompress.js"));
  const input = JSON.parse(fs.readFileSync(0, "utf8"));
  if ((input.tool_name || input.toolName) !== "Bash") passthrough();

  const resp = input.tool_response ?? input.toolResponse ?? input.tool_result;
  // Claude's Bash PostToolUse result is an object. updatedToolOutput must retain
  // that object shape; returning only the replacement string is rejected by the
  // hook runtime and causes a generic hook error.
  if (!resp || typeof resp !== "object" || Array.isArray(resp)) passthrough();
  const textField = ["stdout", "output", "content"]
    .find((field) => typeof resp[field] === "string");
  if (!textField) passthrough();
  const text = resp[textField];
  if (typeof text !== "string" || !text) passthrough();

  const { view, saved, dropped } = compress(text);
  // emit when lines collapsed OR ANSI stripping alone saved significant chars
  if (dropped < 1 && saved <= 200) passthrough();

  // stash original so per-line detail is recoverable
  const cacheDir = process.env.XDG_CACHE_HOME
    ? path.join(process.env.XDG_CACHE_HOME, "claude-logcompress")
    : path.join(os.homedir(), ".cache", "claude-logcompress");
  const hash = crypto.createHash("sha256").update(text).digest("hex").slice(0, 16);
  const cachePath = path.join(cacheDir, `${hash}.json`);
  const cacheTmp = `${cachePath}.${process.pid}.tmp`;
  fs.mkdirSync(cacheDir, { recursive: true, mode: 0o700 });
  try {
    fs.writeFileSync(cacheTmp, text, { mode: 0o600 });
    fs.renameSync(cacheTmp, cachePath);
  } finally {
    fs.rmSync(cacheTmp, { force: true });
  }

  const parts = [];
  if (dropped > 0) {
    parts.push(`collapsed ${dropped} repeated line(s)`);
    if (saved > 0) parts.push(`saved ${saved} chars total`);
  } else if (saved > 0) {
    parts.push(`stripped ${saved} ANSI chars`);
  }
  const note = `\n[logcompress: ${parts.join(", ")}. Full output: cat ${cacheDir}/${hash}.json]`;
  emit({
    hookSpecificOutput: {
      hookEventName: "PostToolUse",
      updatedToolOutput: { ...resp, [textField]: view + note },
    },
  });
} catch {
  passthrough(); // fail open — never corrupt a tool result
}

File: profiles/tom/hooks/logcompress-hook.py
#!/usr/bin/env python3
# Adapted from https://github.com/Green-PT/honey-for-devs hooks/ (MIT).
# Copyright (c) 2026 Green-PT
"""Fail-open Codex PostToolUse compression hook."""

from __future__ import annotations

import hashlib
import json
import os
from pathlib import Path
import select
import subprocess
import sys
import time


HERE = Path(__file__).resolve().parent
MAX_INPUT_BYTES = 64 * 1024 * 1024
READ_TIMEOUT_SECONDS = 1.0
INNER_TIMEOUT_SECONDS = 2.0


def emit(payload: object) -> None:
    sys.stdout.write(json.dumps(payload, ensure_ascii=False, separators=(",", ":")))


def read_bounded_input() -> bytes | None:
    """Drain stdin completely, or return None at the size/deadline boundary."""
    deadline = time.monotonic() + READ_TIMEOUT_SECONDS
    chunks: list[bytes] = []
    size = 0
    while True:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            return None
        ready, _, _ = select.select([sys.stdin.fileno()], [], [], remaining)
        if not ready:
            return None
        chunk = os.read(sys.stdin.fileno(), 64 * 1024)
        if not chunk:
            return b"".join(chunks)
        size += len(chunk)
        if size > MAX_INPUT_BYTES:
            return None
        chunks.append(chunk)


def hook_enabled() -> bool:
    environment = dict(os.environ)
    environment["NORTH_AGENT_PYTHON"] = sys.executable
    try:
        result = subprocess.run(
            [
                "bash",
                "-c",
                'source "$1" && north_hook_enabled "$2"',
                "logcompress-dial",
                str(HERE / "lib" / "harness-dial.sh"),
                "logcompress-hook",
            ],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            env=environment,
            timeout=1.0,
            check=False,
        )
    except (OSError, subprocess.SubprocessError):
        return True
    return result.returncode != 1


def supervise() -> None:
    payload = read_bounded_input()
    if payload is None or not hook_enabled():
        return
    environment = dict(os.environ)
    environment["LOGCOMPRESS_INNER"] = "1"
    try:
        child = subprocess.run(
            [sys.executable, str(Path(__file__).resolve())],
            input=payload,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            env=environment,
            timeout=INNER_TIMEOUT_SECONDS,
            check=False,
        )
        if child.returncode != 0 or not child.stdout:
            return
        result = json.loads(child.stdout)
        output = result.get("hookSpecificOutput")
        if isinstance(output, dict) and output.get("hookEventName") == "PostToolUse":
            sys.stdout.buffer.write(child.stdout)
    except (OSError, ValueError, subprocess.SubprocessError):
        return


def inner() -> None:
    try:
        if os.environ.get("LOGCOMPRESS_TEST_HANG") == "1":
            pid_file = os.environ.get("LOGCOMPRESS_TEST_PID_FILE")
            if pid_file:
                Path(pid_file).write_text(str(os.getpid()), encoding="utf-8")
            time.sleep(30)

        from logcompress import compress

        payload = json.load(sys.stdin)
        if (payload.get("tool_name") or payload.get("toolName")) != "Bash":
            return
        response = payload.get("tool_response")
        if response is None:
            response = payload.get("toolResponse")
        if response is None:
            response = payload.get("tool_result")
        if not isinstance(response, dict):
            return

        text_field = next(
            (
                field
                for field in ("stdout", "output", "content")
                if isinstance(response.get(field), str)
            ),
            None,
        )
        if text_field is None or not response[text_field]:
            return
        text = response[text_field]
        view, saved, dropped = compress(text)
        if dropped < 1 and saved <= 200:
            return

        cache_root = os.environ.get("XDG_CACHE_HOME")
        if not cache_root:
            cache_root = str(Path.home() / ".cache")
        cache_dir = Path(cache_root) / "claude-logcompress"
        digest = hashlib.sha256(text.encode()).hexdigest()[:16]
        cache_path = cache_dir / f"{digest}.json"
        cache_tmp = cache_dir / f"{digest}.json.{os.getpid()}.tmp"
        cache_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
        try:
            descriptor = os.open(
                cache_tmp, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600
            )
            with os.fdopen(descriptor, "w", encoding="utf-8") as handle:
                handle.write(text)
            os.replace(cache_tmp, cache_path)
        finally:
            try:
                cache_tmp.unlink()
            except FileNotFoundError:
                pass

        parts: list[str] = []
        if dropped > 0:
            parts.append(f"collapsed {dropped} repeated line(s)")
            if saved > 0:
                parts.append(f"saved {saved} chars total")
        elif saved > 0:
            parts.append(f"stripped {saved} ANSI chars")
        note = (
            f"\n[logcompress: {', '.join(parts)}. "
            f"Full output: cat {cache_dir}/{digest}.json]"
        )
        replacement = dict(response)
        replacement[text_field] = view + note
        emit(
            {
                "hookSpecificOutput": {
                    "hookEventName": "PostToolUse",
                    "updatedToolOutput": replacement,
                }
            }
        )
    except Exception:
        return


if __name__ == "__main__":
    try:
        if os.environ.get("LOGCOMPRESS_INNER") == "1":
            inner()
        else:
            supervise()
    except Exception:
        pass
